### ISSUE TITLE
Fatal Error Webdav on Windows10 in NC15

### ISSUE THREAD

**Issue by Bables55** [nextcloud/server#13043]
### Steps to reproduce
1. Mount nextcloud on windows10 by Network drive
2. Open target file.
3.

### Expected behaviour
There is no problem.

### Actual behaviour
It seems that it can be operated normally in appearance.
However, when I look at the log of Nextcloud15 when the user operates it is always recorded an error.
This did not occur in NC14.

### Server configuration

**Operating system**:
CentOS 7.5
**Web server:**
nginx 1.13.8
**Database:**
MariaDB 10.2.19
**PHP version:**
7.1.24
**Nextcloud version:** (see Nextcloud admin page)
15.0.0
**Updated from an older Nextcloud/ownCloud or fresh install:**
14.0.4
**Where did you install Nextcloud from:**
get source from nextcloud.com
**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - admin_audit: 1.5.0
  - announcementcenter: 3.4.0
  - apporder: 0.6.0
  - calendar: 1.6.4
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 2.1.8
  - dav: 1.8.0
  - deck: 0.5.1
  - encryption: 2.3.0
  - end_to_end_encryption: 1.1.0
  - external: 3.2.0
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_accesscontrol: 1.5.0
  - files_antivirus: 2.0.0
  - files_automatedtagging: 1.5.0
  - files_external: 1.6.0
  - files_markdown: 2.0.5
  - files_pdfviewer: 1.4.0
  - files_retention: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - gallery: 18.2.0
  - groupfolders: 2.0.1
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - mail: 0.11.0
  - nextcloud_announcements: 1.4.0
  - notes: 2.5.1
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - provisioning_api: 1.5.0
  - ransomware_protection: 1.3.0
  - richdocuments: 3.0.5
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - spreed: 4.99.0
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - tasks: 0.9.8
  - theming: 1.6.0
  - theming_customcss: 1.2.0
  - twofactor_backupcodes: 1.4.1
  - updatenotification: 1.5.0
  - user_usage_report: 1.1.1
  - workflowengine: 1.5.0
Disabled:
  - drawio
  - firstrunwizard
  - mindmaps
  - quicknotes
  - sharepoint
  - user_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
"system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "XXXXXXXXXX",
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/XXXXXXXXXX",
        "dbtype": "mysql",
        "version": "15.0.0.10",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "log_type": "file",
        "logtimezone": "Asia\/Tokyo",
        "loglevel": 2,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": true,
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpsecure": "ssl",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "auth.bruteforce.protection.enabled": false,
        "versions_retention_obligation": "7,auto",
        "trashbin_retention_obligation": "auto",
        "defaultapp": "calendar",
        "theme": "",
        "updater.release.channel": "stable",
        "twofactor_enforced": "false",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": []
    }
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
webdav
**Are you using encryption:** yes/no
no
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

### Client configuration
**Browser:**
Firefox, Vivaldi

**Operating system:**
Windows10 (1809)

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
[webdav] Fatal: Sabre\DAV\Exception: Error while copying file to target location (copied bytes: 0, expected filesize: 0 ) at <<closure>>

0. /MYNEXTCLOUD/apps/dav/lib/Connector/Sabre/Directory.php line 156
   put(null)
1. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 1096
   createFile("~$FILENAME.xlsx", null)
2. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 525
   createFile("NC-USERNAME/DIRECTORY/~$FILENAME.xlsx", null, null)
3. <<closure>>
   httpPut(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
4. /MYNEXTCLOUD/3rdparty/sabre/event/lib/EventEmitterTrait.php line 105
   call_user_func_array([Sabre\DAV\CorePlugin {},"httpPut"], [Sabre\HTTP\Requ ... }])
5. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 479
   emit("method:PUT", [Sabre\HTTP\Requ ... }])
6. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 254
   invokeMethod(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
7. /MYNEXTCLOUD/apps/dav/appinfo/v1/webdav.php line 80
   exec()
8. /MYNEXTCLOUD/remote.php line 163
   require_once("/MYNEXTCLO ... p")

PUT /remote.php/webdav/NC-USERNAME/DIRECTORY/~%FILENAME.xlsx
from USER-IPADDRESS by NC-USERNAME at 2018-12-13T10:17:05+09:00

[no app in context] Error: Sabre\DAV\Exception: Error while copying file to target location (copied bytes: 0, expected filesize: 0 ) at <<closure>>

0. /MYNEXTCLOUD/apps/dav/lib/Connector/Sabre/Directory.php line 156
   put(null)
1. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 1096
   createFile("~$FILENAME.xlsx", null)
2. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 525
   createFile("NC-USERNAME/DIRECTORY/~$FILENAME.xlsx", null, null)
3. <<closure>>
   httpPut(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
4. /MYNEXTCLOUD/3rdparty/sabre/event/lib/EventEmitterTrait.php line 105
   call_user_func_array([Sabre\DAV\CorePlugin {},"httpPut"], [Sabre\HTTP\Requ ... }])
5. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 479
   emit("method:PUT", [Sabre\HTTP\Requ ... }])
6. /MYNEXTCLOUD/3rdparty/sabre/dav/lib/DAV/Server.php line 254
   invokeMethod(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
7. /MYNEXTCLOUD/apps/dav/appinfo/v1/webdav.php line 80
   exec()
8. /MYNEXTCLOUD/remote.php line 163
   require_once("/MYNEXTCLO ... p")

PUT /remote.php/webdav/NC-USERNAME/DIRECTORY/~%FILENAME.xlsx
from USER-IPADDRESS by NC-USERNAME at 2018-12-13T10:17:05+09:00

[PHP] Error: feof(): supplied resource is not a valid stream resource at /MYNEXTCLOUD/apps/dav/lib/Connector/Sabre/File.php#172

PUT /remote.php/webdav/NC-USERNAME/DIRECTORY/~%FILENAME.xlsx
from USER-IPADDRESS by NC-USERNAME at 2018-12-13T10:17:05+09:00
```
</details>

**Comment by algbaer** [comment]
Same problem here with NC 15. It works after disabling files_antivirus app. Actually it does not seem to be webdav related. You also can't upload a file with 0 byte size via web UI when files_antivirus is active.

**Comment by Bables55** [comment]
It seems to be caused by "Antivirus for files" and "Default encrypt module (Encryption)". I can not upload files on WebDAV on Mac and Windows if I activate either.
When I posted this issue, it was the same environment. At that time the error log remained but the upload itself was possible. Is it a timed problem?

**Comment by Schnuckel65** [comment]
The same problem at me. I cannot upload any file via WebDAV on Windows since updated to NC15. The upload stops in the Windows Explorer with the error message "Die Quelldatei oder vom Quelldatenträger kann nicht gelesen werden." in my German Windows. The log than contains a lot of Sabre\DAV\Exception entries which starts with "Symfony\Component\Routing\Exception\RouteNotFoundException: Unable to generate a URL for the named route "ocsms.sms.index" as such route does not exist.".
Deactivating the "Antivirus for files" or the "Default encryption module" does not help in my case.

**Comment by j-ed** [comment]
Due to the fact that a named route "ocsms.sms.index" is missing, you should try to deactivate the ["ocsms" app](https://github.com/nextcloud/ocsms) (Nextcloud/ownCloud PhoneSync server application) first. An other possible solution might be https://projects.duckcorp.org/issues/444 .

**Comment by algbaer** [comment]
Hoped that 15.0.5 with #14273 fixes this, but enabling files_antivirus still breaks uploads of 0 byte files.

**Comment by Schnuckel65** [comment]
Great job! It works. Thank's a lot.

**Comment by Bables55** [comment]
Thank you.
I updated to NC 15.0.5 and enabled Antivirus.app(2.0.1) for a long time. Then I did not get an error. And I'm uploading on Windows.
*I have updated PHP to 7.3.2 from that.

**Comment by schmidtadrian** [comment]
I'm still having this error with NC 15.0.5 and Antivirus 2.0.1

**Comment by Bables55** [comment]
If I change ‘/var/run/clamd.scan’ permission to 777, I get the same WebDAV problem as before.
I can avoid the problem that can not be uploaded by setting its permissions to 700 or 770. However, in this case, there is a problem with permission to clmad.sock. Error 13.
In my case, the permissions were 700 and WebDAV problems were occurring.

**Comment by schmidtadrian** [comment]
If I change the Permission as you said i also get Error 13 (Permission denied).
But if change the Permission back get the no app in context error again

**Comment by Bables55** [comment]
I do not know the correct permissions of the clamd.scan directory and clamd.sock. Which one is correct?

It is commonly said that there are many problems with WebDAV built in Windows, so please use a third party (WebDAV) client. There is also this, so I have not been able to judge what is right. For my convenience I would like to use Windows built-in WebDAV.

I no longer get the error "the no app in context error". Basically two errors below occurred at the same time in my error. The combination is displayed a plurality of times. *clamd.scan directory is 700, clamd.scan is 666. 

```
[files_antivirus] Error: RuntimeException: Cannot connect to "/var/run/clamd.scan/clamd.sock": Permission denied (code 13) at <<closure>>

 0. /Nextcloud Directory/apps/files_antivirus/lib/AvirWrapper.php line 108
    initScanner()
 1. /Nextcloud Directory/apps/files_antivirus/lib/AvirWrapper.php line 92
    wrapSteam("files/NC-USER/FILENAM ... t", null)
 2. /Nextcloud Directory/lib/private/Files/Storage/Wrapper/Wrapper.php line 630
    writeStream("files/NC-USER/FILENAM ... t", null, null)
 3. /Nextcloud Directory/lib/private/Files/Storage/Wrapper/Wrapper.php line 630
    writeStream("files/NC-USER/FILENAM ... t", null, null)
 4. /Nextcloud Directory/lib/private/Files/Storage/Wrapper/Wrapper.php line 630
    writeStream("files/NC-USER/FILENAM ... t", null, null)
 5. /Nextcloud Directory/apps/dav/lib/Connector/Sabre/File.php line 182
    writeStream("files/NC-USER/FILENAM ... t", null)
 6. /Nextcloud Directory/apps/dav/lib/Connector/Sabre/Directory.php line 156
    put(null)
 7. /Nextcloud Directory/3rdparty/sabre/dav/lib/DAV/Server.php line 1096
    createFile("FILENAME.xlsx", null)
 8. /Nextcloud Directory/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 525
    createFile("NC-USER/FILENAME.xlsx", null, null)
 9. <<closure>>
    httpPut(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
10. /Nextcloud Directory/3rdparty/sabre/event/lib/EventEmitterTrait.php line 105
    call_user_func_array([Sabre\DAV\CorePlugin {},"httpPut"], [Sabre\HTTP\Requ ... }])
11. /Nextcloud Directory/3rdparty/sabre/dav/lib/DAV/Server.php line 479
    emit("method:PUT", [Sabre\HTTP\Requ ... }])
12. /Nextcloud Directory/3rdparty/sabre/dav/lib/DAV/Server.php line 254
    invokeMethod(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
13. /Nextcloud Directory/apps/dav/appinfo/v1/webdav.php line 80
    exec()
14. /Nextcloud Directory/remote.php line 163
    require_once("/Nextcloud Dire ... p")

PUT /remote.php/webdav/NC-USER/FILENAME.xlsx
```
```

[PHP] Error: stream_socket_client(): unable to connect to unix:///var/run/clamd.scan/clamd.sock (Permission denied) at /Nextcloud Directory/apps/files_antivirus/lib/Scanner/External.php#41

PUT /remote.php/webdav/NC-USER/FILENAME.xlsx

```

**Comment by j-ed** [comment]
> The combination is displayed a plurality of times. *clamd.scan directory is 700, clamd.scan is 666.

How should it work, if the directory access rights are **0700**? The "7" gives full access for the owner,
the next "0" says that **no group** access has been given and the last "0" says that **no world** access has
been given either. Check-out [**this issue**](https://github.com/nextcloud/files_antivirus/issues/84#issuecomment-388318821) which explains how the problem has been fixed on CentOS.

**Comment by algbaer** [comment]
> It is commonly said that there are many problems with WebDAV built in Windows, so please use a third party (WebDAV) client. There is also this, so I have not been able to judge what is right. For my convenience I would like to use Windows built-in WebDAV.
> 
The Windows WebDAV client uploads in two steps: 1. creates an empty file, 2. puts the file with contents in place.

With files_antivirus enabled, actions to trigger the problem:

- Uploading empty file via web UI
- Doing WebDAV PUT of empty file (or using Windows WebDAV client with any file, because of the above mentioned behavior)
- Copy an empty file to Nextcloud desktop client sync directory

@Bables55: You can test proper files_antivirus setup by uploading the eicar anti malware test file.

**Comment by algbaer** [comment]
Fixed for me in 15.0.6, see #14517
